### PR TITLE
Add helper script for Black Magic Probe

### PR DESCRIPTION
**Pull request by maz3max**
This script can detect [Black Magic Probes](https://github.com/blacksphere/blackmagic/wiki) and act as a flashloader (and more).
It is compatible with Linux and macOS.

All important options that can be set via the monitor command are available as arguments.

Makefiles for using `make flash`, `make debug` and `make term` with the script are included.

### Contribution description
Involved parts: build system
Kind of contribution: support for new kind of debug probe

### Testing procedure
add something like this to your boards `Makefile.include` (tested with bluepill-like board):
```
BMP_OPTIONS = --connect-srst
FLASHFILE ?= $(ELFFILE)
include $(RIOTMAKE)/tools/bmp.inc.mk
include $(RIOTMAKE)/tools/bmp-serial.inc.mk
```
and use make flash etc.

### CODE REVIEW

**Comment by maz3max** [comment]
Note: This currently uses gdb-multiarch and picocom. Different GDB and TTY programs can be specified in BMP_OPTIONS.

**Comment by maz3max** [comment]
phew - that was more of a struggle than I thought.
Sorry for the numerous commits - should I squash them into one?

**Comment by benpicco** [comment]
As nobody has reviewed this yet, I'd say go ahead and squash.
This is a handy addition, but I don't know enough Python to confidently review this.

File: dist/tools/bmp/bmp.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
+#!/usr/bin/python3
+# Black Magic Probe helper script
+# This script can detect connected Black Magic Probes and can be used as a flashloader and much more
+
```
Comment (miri64) [inline dist/tools/bmp/bmp.py]: License note missing.

File: dist/tools/bmp/bmp.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                if msg['message'] == expected_result:
+                    print(description, "successful.")
+                    return True
+                else:
+                    print(description, "failed.", file=sys.stderr)
+                    return False
+        res = gdbmi.get_gdb_response(timeout_sec=TIMEOUT)
+
+
+if __name__ == '__main__':
```
Comment (miri64) [inline dist/tools/bmp/bmp.py]: At least some of the code below could go into its own function for easier readability.

Comment (maz3max) [reply dist/tools/bmp/bmp.py]: Do you mean code from the main function?

File: dist/tools/bmp/bmp.py (hunk 1 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: Any suggestions for usable licenses?

Comment (maz3max) [reply dist/tools/bmp/bmp.py]: Added a license.

File: dist/tools/bmp/bmp.py (hunk 2 above)
Comment (miri64) [reply dist/tools/bmp/bmp.py]: Well. there is no main function ;-). But yes, moving some of the stuff below into their own functions definitely helped with readability.

Comment (maz3max) [reply dist/tools/bmp/bmp.py]: Yeah, you're right. ^^' What do think about the code? Is it readable now?

**Comment by maribu** [comment]
Works like a charm. I'd be super happy to see BMP support in RIOT.

@maz3max: One minor feature request (feel free to ignore): If --gdb-path is not explicitly given, maybe the script could try to use both `arm-none-eabi-gdb` and `gdb-multiarch` for the GDB program name?

@miri64: I think your comments have been resolved, is that correct?

**Comment by maribu** [comment]
BTW: A BluePill can be converted to a BlackMagic Probe by following theses instructions: https://medium.com/@paramaggarwal/converting-an-stm32f103-board-to-a-black-magic-probe-c013cf2cc38c#.btn6lnwqe

@MrKevinWeiss: I heard you are the BluePill pope?

**Comment by maribu** [comment]
Btw, I tested with a ST-Link v2 clone converted to a BMP and flashed a BluePill.

<details><summary>Nasty hack to add BMP support to the bluepill board</summary>

This wouldn't apply to RIOTs current master any more. But this a quick and dirty patch anyway.

``` diff
diff --git a/boards/common/stm32f103c8/Makefile.include b/boards/common/stm32f103c8/Makefile.include
index 9862d252c8..51981f265c 100644
--- a/boards/common/stm32f103c8/Makefile.include
+++ b/boards/common/stm32f103c8/Makefile.include
@@ -10,11 +10,8 @@ endif
 INCLUDES += -I$(RIOTBOARD)/common/stm32f103c8/include
 
 # define the default port depending on the host OS
-PORT_LINUX ?= /dev/ttyUSB0
 PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.SLAB_USBtoUART*)))
 
-# setup serial terminal
-include $(RIOTMAKE)/tools/serial.inc.mk
 
 # optionally, use dfu-util to flash via usb
 # note: needs a bootloader flashed before, config below is compatible
@@ -33,13 +30,8 @@ ifeq ($(PROGRAMMER),dfu-util)
   # FFLAGS = -d 1d50:6017 -s 0x08002000:leave -D $(FLASHFILE)
 else
 
-  # this board uses openocd by default
-  DEBUG_ADAPTER ?= stlink
-  STLINK_VERSION ?= 2
-
-  # call a 'reset halt' command before starting the debugger
-  # it is required as `connect_assert_srst` is set
-  export OPENOCD_DBG_START_CMD = -c 'reset halt'
-
-  include $(RIOTMAKE)/tools/openocd.inc.mk
+  BMP_OPTIONS = --connect-srst --gdb-path arm-none-eabi-gdb
+  FLASHFILE ?= $(ELFFILE)
+  include $(RIOTMAKE)/tools/bmp.inc.mk
+  include $(RIOTMAKE)/tools/bmp-serial.inc.mk
 endif
```
</details>

<details><summary>Output of make flash</summary>
```
$ make flash
Building application "default" for "bluepill" with MCU "stm32f1".

"make" -C /home/maribu/Repos/software/RIOT/boards/bluepill
"make" -C /home/maribu/Repos/software/RIOT/boards/common/stm32f103c8
"make" -C /home/maribu/Repos/software/RIOT/core
"make" -C /home/maribu/Repos/software/RIOT/cpu/stm32f1
"make" -C /home/maribu/Repos/software/RIOT/cpu/cortexm_common
"make" -C /home/maribu/Repos/software/RIOT/cpu/cortexm_common/periph
"make" -C /home/maribu/Repos/software/RIOT/cpu/stm32_common
"make" -C /home/maribu/Repos/software/RIOT/cpu/stm32_common/periph
"make" -C /home/maribu/Repos/software/RIOT/cpu/stm32f1/periph
"make" -C /home/maribu/Repos/software/RIOT/drivers
"make" -C /home/maribu/Repos/software/RIOT/drivers/periph_common
"make" -C /home/maribu/Repos/software/RIOT/drivers/saul
"make" -C /home/maribu/Repos/software/RIOT/sys
"make" -C /home/maribu/Repos/software/RIOT/sys/auto_init
"make" -C /home/maribu/Repos/software/RIOT/sys/auto_init/saul
"make" -C /home/maribu/Repos/software/RIOT/sys/fmt
"make" -C /home/maribu/Repos/software/RIOT/sys/isrpipe
"make" -C /home/maribu/Repos/software/RIOT/sys/newlib_syscalls_default
"make" -C /home/maribu/Repos/software/RIOT/sys/phydat
"make" -C /home/maribu/Repos/software/RIOT/sys/pm_layered
"make" -C /home/maribu/Repos/software/RIOT/sys/ps
"make" -C /home/maribu/Repos/software/RIOT/sys/saul_reg
"make" -C /home/maribu/Repos/software/RIOT/sys/shell
"make" -C /home/maribu/Repos/software/RIOT/sys/shell/commands
"make" -C /home/maribu/Repos/software/RIOT/sys/stdio_uart
"make" -C /home/maribu/Repos/software/RIOT/sys/tsrb
   text	   data	    bss	    dec	    hex	filename
  18732	    504	   2652	  21888	   5580	/home/maribu/Repos/software/RIOT/examples/default/bin/bluepill/default.elf
/home/maribu/Repos/software/RIOT/dist/tools/bmp/bmp.py --connect-srst --gdb-path arm-none-eabi-gdb flash /home/maribu/Repos/software/RIOT/examples/default/bin/bluepill/default.elf
found following Black Magic GDB servers:
	[/dev/ttyACM0] Serial: C1B5BFDA <- default 
connecting to [/dev/ttyACM0]...
connecting successful.
scanning using SWD...
found following targets:
	STM32F1 medium density

attaching to target successful.
downloading... total size: 902.6K
downloading section [.text] (18.3K)
100%|#################################################################################################################################################################################################################################|
downloading section [.relocate] (504B)
100%|#################################################################################################################################################################################################################################|
downloading finished
checking successful
killing successful.
```
</details>

<details><summary>Output of make term</summary>

```
$ make term
/home/maribu/Repos/software/RIOT/dist/tools/bmp/bmp.py --connect-srst --gdb-path arm-none-eabi-gdb term
picocom v3.1

port is        : /dev/ttyACM1
flowcontrol    : none
baudrate is    : 115200
parity is      : none
databits are   : 8
stopbits are   : 1
escape is      : C-a
local echo is  : no
noinit is      : no
noreset is     : no
hangup is      : no
nolock is      : yes
send_cmd is    : sz -vv
receive_cmd is : rz -vv -E
imap is        : lfcrlf,
omap is        : 
emap is        : crcrlf,delbs,
logfile is     : none
initstring     : none
exit_after is  : not set
exit is        : no

Type [C-a] [C-h] to see available commands
Terminal ready
help
Command              Description
---------------------------------------
reboot               Reboot the node
ps                   Prints information about running threads.
saul                 interact with sensors and actuators using SAUL
> ps
	pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
	  - | isr_stack            | -        - |   - |    512 (  116) | 0x20000000 | 0x200001c8
	  1 | idle                 | pending  Q |  15 |    256 (  128) | 0x20000534 | 0x200005b4 
	  2 | main                 | running  Q |   7 |   1536 (  648) | 0x20000634 | 0x20000aa4 
	    | SUM                  |            |     |   2304 (  892)
>
```

</details>

File: dist/tools/bmp/bmp.py (hunk 2 above)
Comment (maribu) [reply dist/tools/bmp/bmp.py]: @miri64: Ping?

Comment (miri64) [reply dist/tools/bmp/bmp.py]: Better :-)

**Comment by maribu** [comment]
OK. If I see this correctly, @miri64's comments have been resolved, right? So if she agrees, I'd say squash. I'll test one final time to be sure and then I think this should be ready for merging, if no one objects.

(I've been using black magic probes (with custom hardware) for quite some time, and it worked much more reliable than the stlinks for me. Would be super nice to be able to use `make DEBUGGER=bmp debug` for me :-))

**Comment by maz3max** [comment]
> > miri64 14 minutes ago
> > Better :-)
> 
> OK, then please squash!

Squash completed.

File: dist/tools/bmp/bmp.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            print('')
+
+        port = g[0].device
+        if args.port:
+            port = args.port
+        elif args.serial:
+            port = search_serial(args.serial, g)
+            assert port, "no BMP with this serial found"
+
+        print('connecting to [%s]...' % port)
```
Comment (maribu) [inline dist/tools/bmp/bmp.py]: Maybe this could be split out into another helper function, e.g. `find_bmp()`?

File: dist/tools/bmp/bmp.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+    # terminal mode, opens TTY program
+    if args.action == 'term':
+        port = u[0].device
+        if args.port:
+            port = args.port
+        elif args.serial:
+            port = search_serial(args.serial, u)
+            assert port, "no BMP with this serial found"
+        os.system(args.term_cmd % port)
+        sys.exit(0)
```
Comment (maribu) [inline dist/tools/bmp/bmp.py]: How about having the individual actions as separate functions, e.g. something like `term_mode()`?

File: dist/tools/bmp/bmp.py [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+            port = search_serial(args.serial, g)
+            assert port, "no BMP with this serial found"
+
+        print('connecting to [%s]...' % port)
+
+        args.file = args.file if args.file else ''
+        args.gdb_path = find_suitable_gdb(args.gdb_path)
+        if not args.gdb_path:
+            print("CANNOT LOCATE SUITABLE GDB EXECUTABLE!", file=sys.stderr)
+            sys.exit(-1)
```
Comment (maribu) [inline dist/tools/bmp/bmp.py]: How about splitting out as function, e.g. `find_gdb_executable()`?

File: dist/tools/bmp/bmp.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                gdb_args.append('-ex \'monitor tpwr enable\'')
+            if args.connect_srst:
+                gdb_args.append('-ex \'monitor connect_srst enable\'')
+            if args.jtag:
+                gdb_args.append('-ex \'monitor jtag_scan\'')
+            else:
+                gdb_args.append('-ex \'monitor swdp_scan\'')
+            gdb_args.append('-ex \'attach %s\'' % args.attach)
+            os.system(" ".join(['\"' + args.gdb_path + '\"'] + gdb_args + [args.file]))
+            sys.exit(0)
```
Comment (maribu) [inline dist/tools/bmp/bmp.py]: How about a separate function, e.g. `debug_mode()`

File: dist/tools/bmp/bmp.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            res = gdbmi.write('monitor swdp_scan', timeout_sec=TIMEOUT)
+        else:
+            print("scanning using JTAG...")
+            res = gdbmi.write('monitor jtag_scan', timeout_sec=TIMEOUT)
+        targets = detect_targets(gdbmi, res)
+        assert len(targets) > 0, "no targets found"
+        print("found following targets:")
+        for t in targets:
+            print("\t%s" % t)
+        print("")
```
Comment (maribu) [inline dist/tools/bmp/bmp.py]: How about a separate function for this, e.g. `connect_to_bmp()`?

**Comment by maribu** [comment]
I have to (and will) test this again

File: dist/tools/bmp/bmp.py (hunk 7 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: this part became `connect_to_target`

File: dist/tools/bmp/bmp.py (hunk 6 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: this part became debug_mode(port)

File: dist/tools/bmp/bmp.py (hunk 5 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: I moved 211-213 into the `find_suitable_gdb(gdb_path)` function

**Comment by maribu** [comment]
Please squash

File: dist/tools/bmp/bmp.py (hunk 4 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: this part became `term_mode(uart_ports)`

File: dist/tools/bmp/bmp.py (hunk 3 above)
Comment (maz3max) [reply dist/tools/bmp/bmp.py]: this part became `choose_bmp_port(gdb_ports)`

**Comment by maz3max** [comment]
squashed

**Comment by miri64** [comment]
Fine by me as well :-)

**Comment by Citrullin** [comment]
> BTW: A BluePill can be converted to a BlackMagic Probe by following theses instructions: https://medium.com/@paramaggarwal/converting-an-stm32f103-board-to-a-black-magic-probe-c013cf2cc38c#.btn6lnwqe
> 
> @MrKevinWeiss: I heard you are the BluePill pope?

Didn't work for me. The article is outdated and the project changed a lot. This article works fine: https://buger.dread.cz/black-magic-probe-bmp-on-bluepill-stm32f103c8-minimum-system-development-board.html
[Just flashed 11 bluepills/blackpills over the weekend. ](https://twitter.com/Citrullin/status/1185490869551599617) :D

Huge thanks to @maz3max for these scripts. They help a lot. :) I just used the scripts standalone, without the make implementation.
Would be amazing to have it available with PROGAMMER=bmp and DEBUGGER=bmp

**Pull request merged**